label_studio/projects/models.py

```diff
@@ -843,8 +843,7 @@ def _update_tasks_counters(self, queryset, from_scratch=True):
             task.total_predictions = task.new_total_predictions
             objs.append(task)
 
-        with transaction.atomic():
-            bulk_update(objs, update_fields=['total_annotations', 'cancelled_annotations', 'total_predictions'], batch_size=settings.BATCH_SIZE)
+        bulk_update(objs, update_fields=['total_annotations', 'cancelled_annotations', 'total_predictions'], batch_size=settings.BATCH_SIZE)
```

Comment:
My understanding is that one of the primary goals of the transaction approach was to ensure that if the tasks' counters get an update, so should the `is_enabled` fields on the tasks. This is why I took a similar approach in writing https://github.com/heartexlabs/label-studio/pull/4511/files#diff-6ff9748c9ae68f8cccd65f4ee20df06ed0a8b9c32b58da39db3001a7178b1dabL861-L866 - but it seems like this code is still too lock-prone in practice. Under some conditions `bulk_update_stats_project_tasks` will make 2000+ sql queries https://github.com/heartexlabs/label-studio/pull/4511/files#diff-278a5eaed0e213875745abafc04f21d5564f53bbcf2160a6469d1ea0e3526e24R892-R893, while holding a lock on all the rows affected by `_update_tasks_counters`

Reply:
...but then, there is the possibility in the code as written that instead an OperationalError exception doesn't propagate up to where the atomic context manager can see it, and therefore, despite an error, we still don't roll back the is_enabled changes: https://github.com/heartexlabs/label-studio/pull/4511/files#diff-278a5eaed0e213875745abafc04f21d5564f53bbcf2160a6469d1ea0e3526e24R897

Reply:
Is this currently synchronous in the UI? Wonder if we can move this to be handled by rq workers where potential retries wouldn't be an issue and it could be done without atomic as it is being done here.  We could then mark `is_enabled` when it is complete.  Just a thought; I haven't looked at the entirety of related code so speaking partially from ignorance.

Reply:
As far as I can tell, no, this code doesn't get called synchronously anywhere; it's only executed on workers as-is.